---
title: "CRAFT functional predictions"
subtitle: "Overview of circRNA functional predictions for miRNA and RBP binding sites, and coding potential"
author: "Please cite: _Dal Molin et al. CRAFT: a bioinformatics software for custom prediction of circular RNA functions_"
date: "`r Sys.Date()`"
header-includes:
    \usepackage{caption}
output:
  rmdformats::readthedown:
  html_document:
    fig_caption: yes
    number_sections: yes
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
  editor_options: 
    chunk_output_type: inline
params:
    l: 50000
    QUANTILE1: "FALSE"
    thr1: 0.95
    score_miRNA: 80
    energy_miRNA: -15
    QUANTILE2: "FALSE"
    thr2: 0.95
    dGduplex_miRNA: -15
    dGopen_miRNA: -15
    QUANTILE3: "FALSE"
    thr3: 0.9
    voteFrac_RBP: 0.15
    orgdb: "org.Hs.eg.db"
    symbol2eg: "org.Hs.egSYMBOL2EG"
    eg2uniprot: "org.Hs.egUNIPROT"
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, dpi = 600)

```

```{r libraries, echo=FALSE, include=FALSE}

install.packages("BiocManager")
library(BiocManager)

BiocManager::install("dplyr")
library(dplyr)
BiocManager::install("tibble")
library(tibble)
BiocManager::install("data.table")
library(data.table)
BiocManager::install("ggplot2")
library(ggplot2)
BiocManager::install("pheatmap")
library(pheatmap)
BiocManager::install("RColorBrewer")
library(RColorBrewer)
BiocManager::install("viridis")
library(viridis)
BiocManager::install("devtools")
library(devtools)
devtools::install_github("kassambara/easyGgplot2")
BiocManager::install("reshape")
library(reshape)
BiocManager::install("reshape2")
library(reshape2)
BiocManager::install("grid")
library(grid)
BiocManager::install("gridExtra")
library(gridExtra)
BiocManager::install("lattice")
library(lattice)
BiocManager::install("DT")
library(DT)
BiocManager::install("gtable")
library(gtable)
BiocManager::install("multiMiR")
library(multiMiR)
BiocManager::install("igraph")
library(igraph)
BiocManager::install("knitr")
library(knitr)
BiocManager::install("rmdformats")
library(rmdformats)

```

```{r input, echo=FALSE, include=FALSE}

## Upload input files

# file of parameters
file_parameters <- "/data/params_R.txt"
parameters <- read.table(file_parameters, header=F)
parameters <- as.data.table(parameters)


if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    ## miRNA prediction
    
    # miRanda
    file_miRanda <- "/data/functional_predictions/miRNA_detection/miRanda/output_miRanda_per_R.txt"
    miRanda_pred <- read.table(file_miRanda, header=T, sep="\t")
    miRanda_pred <- as.data.table(miRanda_pred)

    # PITA
    file_PITA <- "/data/functional_predictions/miRNA_detection/PITA/pred_pita_results_per_R.txt"
    PITA_pred <- read.table(file_PITA, header=T, sep="\t")
    PITA_pred <- as.data.table(PITA_pred)

}


if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    # RBP predictions
    
    file_beRBP <- "/data/functional_predictions/RBP_detection/beRBP/analysis_RBP/resultMatrix_b.txt"
    beRBP_pred <- read.table(file_beRBP, header=T, sep="\t")
    beRBP_pred <- as.data.table(beRBP_pred)

}


if (parameters[1] == "O" | parameters[1] == "MO" | parameters[1] == "RO" | parameters[1] == "MRO") {

    # ORF prediction
    
    file_ORFfinder <- "/data/functional_predictions/ORF_detection/ORFfinder/ORF_backsplice.txt"
    ORFfinder_pred <- read.table(file_ORFfinder, header=T, sep="\t")
    ORFfinder_pred <- as.data.table(ORFfinder_pred)

    file_openORF <- "/data/functional_predictions/ORF_detection/ORFfinder/ORF_backsplice_open.txt"
    openORF <- read.table(file_openORF, header=T, sep="\t")
    openORF <- as.data.table(openORF)

}


# circRNA length
file_circ_length <- "/data/functional_predictions/backsplice_circRNA_length_1.txt"
circ_length <- read.table(file_circ_length, header=T, sep="\t")
circ_length <- as.data.table(circ_length)


# gene names
file_gene_names <- "/data/functional_predictions/backsplice_gene_name.txt"
gene_names <- read.table(file_gene_names, header=T, sep="\t")
gene_names <- as.data.table(gene_names)

```

```{r output, echo=FALSE, include=FALSE}

output_dir <- "/data/graphical_output/general/"

```

```{r, echo=FALSE, include=TRUE}

htmltools::img(src = knitr::image_uri("/input/CRAFT_logo.png"), 
               alt = 'logo', 
               style = 'position:absolute; top: -25px; right: 700px; padding:12px; height:160px; width:160px')

```


```{r, echo=FALSE, include=FALSE}

# remove circRNAs with length > "l"
circ_l <- circ_length[length < params$l, circ_id]

```

Number of circRNAs analyzed: `r length(circ_l)`


```{r info_circ_length, echo=FALSE, include=TRUE}

setkey(gene_names, circ_id)
setkey(circ_length, circ_id)
circ_info <- merge(gene_names, circ_length)


datatable(data = circ_info,
          rownames = F,
          colnames = c("CircRNA", "Gene name", "CircRNA length"),
          style = "bootstrap",
          class = "compact display",
          fillContainer = F,
          autoHideNavigation = T,
          filter = "top",
          escape = FALSE,
          extensions = c('ColReorder', 'Buttons', 'KeyTable'),
          options = list(colReorder = TRUE,
                         searching = TRUE,
                         pageLength = 10,
                         dom = 'Bfrtip',
                         buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                         keys = TRUE
          )
)

```





# MiRNA binding sites prediction

```{r print_0, echo=FALSE, include=FALSE}

print_0 <- "Analysis not performed."

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    print_0 <- ""

}

```

`r print_0`

## Method statistics

```{r, echo=FALSE, include=FALSE}

### miRNA predictions

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    ## miRanda
    
    # remove circRNAs with length > "l"
    miRanda_pred_l <- miRanda_pred[circ_id %in% circ_l,]
    
    
    # filter on "score" ed "energy"
    if (params$QUANTILE1 == "TRUE") {
        score_miRNA <- quantile(miRanda_pred_l$score, params$thr1)
        energy_miRNA <- quantile(miRanda_pred_l$energy, 1-params$thr1)
    } else {
        score_miRNA <- params$score_miRNA
        energy_miRNA <- params$energy_miRNA
    }
    
    miRanda_pred_l_sel <- miRanda_pred_l[(score > score_miRNA & energy < energy_miRNA),]

}

```

```{r, echo=FALSE, include=FALSE}

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    ## PITA
    
    # remove circRNAs with length > "l"
    PITA_pred_l <- PITA_pred[circ_id %in% circ_l,]

    
    # filter on "dGduplex" and "dGopen"
    if (params$QUANTILE2 == "TRUE") {
        dGduplex_miRNA <- quantile(PITA_pred_l$dGduplex, 1-params$thr2)
        dGopen_miRNA <- quantile(PITA_pred_l$dGopen, params$thr2)
    } else {
        dGduplex_miRNA <- params$dGduplex_miRNA
        dGopen_miRNA <- params$dGopen_miRNA
    }
    
    PITA_pred_l_sel <- PITA_pred_l[(dGduplex < dGduplex_miRNA & dGopen > dGopen_miRNA),]

}

```

```{r, echo=FALSE, include=FALSE}

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    # merge predictions of miRanda and PITA
    
    miRanda_pred_l_sel$circ_miRNA_end_id <- paste0(miRanda_pred_l_sel$circ_id, "__", miRanda_pred_l_sel$miRNA_id, "__", miRanda_pred_l_sel$circ_end)
    
    PITA_pred_l_sel$circ_miRNA_end_id <- paste0(PITA_pred_l_sel$circ_id, "__", PITA_pred_l_sel$miRNA_id, "__", PITA_pred_l_sel$circ_end)
    
    
    miRNA_pred <- miRanda_pred_l_sel[which(miRanda_pred_l_sel$circ_miRNA_end_id %in% PITA_pred_l_sel$circ_miRNA_end_id), c("circ_id", "miRNA_id", "circ_end")]
    miRNA_pred$putative_circ_start <- miRNA_pred$circ_end - 22
    miRNA_pred <- data.table("circ_id" = miRNA_pred$circ_id, "miRNA_id" = miRNA_pred$miRNA_id, "putative_circ_start" = miRNA_pred$putative_circ_start, "circ_end" = miRNA_pred$circ_end)

} else {
    
    score_miRNA <- params$score_miRNA
    energy_miRNA <- params$energy_miRNA
    miRanda_pred_l_sel <- data.table()
    
    dGduplex_miRNA <- params$dGduplex_miRNA
    dGopen_miRNA <- params$dGopen_miRNA
    PITA_pred_l_sel <- data.table()

    miRNA_pred <- data.table()
    
}

```

```{r print_1a, echo=FALSE, include=FALSE}

print_1a <- ""

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    print_1a <- paste0("miRanda thresholds: score > ", score_miRNA, ", energy < ", energy_miRNA)

}

```

`r print_1a`

|                                        | miRanda                      |
|----------------------------------------|----------------------------- |
| Number of miRNA binding over threshold | `r miRanda_pred_l_sel[, .N]` |


```{r print_1b, echo=FALSE, include=FALSE}

print_1b <- ""

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    print_1b <- paste0("PITA thresholds: dGduplex < ", dGduplex_miRNA, ", dGopen > ", dGopen_miRNA)
                
}

```

`r print_1b`

|                                        | PITA                      |
|----------------------------------------|-------------------------- |
| Number of miRNA binding over threshold | `r PITA_pred_l_sel[, .N]` |


Number of common miRNA binding sites: `r miRNA_pred[, .N]`.


## MiRNA binding sites per circRNA

```{r pre_barplot_1_miRNA, echo=FALSE, include=FALSE}

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    setkey(miRNA_pred, circ_id)
    setkey(gene_names, circ_id)
    miRNA_pred <- miRNA_pred[gene_names, nomatch = 0]
    
    miRNA_pred$gene_circ_names <- paste0("circ", miRNA_pred$gene_names, "_", miRNA_pred$circ_id)

    bar1a <- miRNA_pred[, .N, by = gene_circ_names]
    
    w <- bar1a[, .N]
    w <- w/2
    if (w < 5) {
        w <- 5
    }

}

```

```{r barplot_1_miRNA, echo=FALSE, include=TRUE}

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {
    
    # barplot of the total number of detected miRNAs in each circRNA
    
    bar_plot1 <- ggplot(bar1a,
                        aes(x = gene_circ_names, y = N, fill = gene_circ_names)
    ) +
        geom_bar(stat = "identity") +
        geom_text(aes(label = N),
                  vjust = -0.3,
                  size = 3.5) +
        scale_fill_viridis(discrete = TRUE) +
        xlab("CircRNA") +
        ylab("Number of miRNA binding sites") +
        ggtitle(paste0("Overall number of miRNA binding sites per circRNA")) +
        theme_bw() +
        theme(plot.title = element_text(hjust = 0.5),
              text = element_text(size=10),
              axis.text.x = element_text(angle=70, hjust=1),
              legend.position = "none")
    
    print(bar_plot1)
    
    # Save at png
    ggsave(bar_plot1, file=paste0(output_dir, "Overall_number_of_miRNA_binding_sites_per_circRNA.png"), width=w, height=7, dpi=600, limitsize = TRUE, type = "cairo")

}

```

```{r datatable_circ_miRNA_length, echo=FALSE, include=TRUE}

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    bar1a$gene_names <- gsub("circ", "", bar1a$gene_circ_names)
    bar1a$gene_names <- gsub("_.*$", "", bar1a$gene_names)
    bar1a$circ_id <- gsub("^.*_", "", bar1a$gene_circ_names)

    bar1a2 <- miRNA_pred[, .(length(unique(miRNA_id))), by = gene_circ_names]
    setkey(bar1a, gene_circ_names)
    setkey(bar1a2, gene_circ_names)
    bar1a <- bar1a[bar1a2, nomatch = 0]
    colnames(bar1a)[5] <- "N_unique"
    
    setkey(bar1a, circ_id)
    setkey(circ_length, circ_id)
    table1a <- merge(bar1a, circ_length)
    table1a <- table1a[, c(1,4,3,5,6)]

    datatable(data = table1a,
          rownames = F,
          colnames = c("CircRNA", "Gene name", "Number of miRNA binding sites", "Number of miRNAs", "CircRNA length"),
          style = "bootstrap",
          class = "compact display",
          fillContainer = F,
          autoHideNavigation = T,
          filter = "top",
          escape = FALSE,
          extensions = c('ColReorder', 'Buttons', 'KeyTable'),
          options = list(colReorder = TRUE,
                         searching = TRUE,
                         pageLength = 10,
                         dom = 'Bfrtip',
                         buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                         keys = TRUE
                    )
        )

}

```

```{r density_miRNA, echo=FALSE, include=FALSE}

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    bar1b <- miRNA_pred[, .N, by = .(circ_id, miRNA_id)]
    
    setkey(bar1b, circ_id)
    setkey(circ_length, circ_id)
    bar1c <- bar1b[circ_length, nomatch = 0]
    
    colnames(bar1c)[3] <- "num_siti"
    
    bar1c <- data.table(bar1c, "density" = bar1c$num_siti/bar1c$length*100)

} else {
    
    miRNA_pred <- data.table(miRNA_id = integer())
    
}

```

Number of different identified miRNAs: `r length(unique(miRNA_pred$miRNA_id))`.


```{r pre_heatmap_num, echo=FALSE, include=FALSE}

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    setkey(bar1b, circ_id)
    setkey(gene_names, circ_id)
    bar1b <- bar1b[gene_names, nomatch = 0]
    
    bar1b$gene_circ_names <- paste0("circ", bar1b$gene_names, "_", bar1b$circ_id)

    if (dim(bar1b)[1] > 0) {
        num_miRNA_per_circRNA <- as.data.table(dcast(bar1b, miRNA_id ~ gene_circ_names, value.var = "N", fill = 0))
    }
    

    # pre heatmaps
    
    # input
    if (dim(bar1b)[1] > 0) {
        
        test <- num_miRNA_per_circRNA[, !"miRNA_id"]
        rownames(test) <- num_miRNA_per_circRNA$miRNA_id
        
        #print only common miRNAs
        test$fun <- apply(test, 1, function(x) length(which(as.integer(x)>0)))
        test <- tibble::rownames_to_column(test, "miRNA_id")
        test <- test[fun >= 2,]
        
        test <- test[, -"fun"]
        test <- tibble::column_to_rownames(test, "miRNA_id")
        
        
        w <- dim(test)[2]/1.67
        if (w < 5) {
            w <- 5
        }
        h <- dim(test)[1]/5
        if (h < 5) {
            h <- 5
        }
    
    }

}

```

```{r canberra_heatmap_num, echo=FALSE, include=FALSE}

# heatmap

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {
    
    if (dim(bar1b)[1] > 0) {
        
        if (dim(test)[1]>=2 & dim(test)[1]<=500 & dim(test)[2]>=2 & dim(test)[2]<=30) {
            
            out <- 0
            i <- 1
            
            while (out == 0) {
                brks <- quantile(test[,i])
                if (length(brks) != length(unique(brks))) {
                    out <- 1
                }
                i <- i+1
            }
            
            if (out == 0) {

                dist = "canberra"
                
                canb_heatmap <- pheatmap(mat = test,
                                         clustering_distance_rows = dist,
                                         clustering_distance_cols = dist,
                                         
                                         color = colorRampPalette(c("white","#1565C0"))(1000),
        
                                         cellheight = 10, cellwidth = 20, border_color="gray",
                                         
                                         treeheight_row = 20, treeheight_col = 20,
        
                                         main = paste0("MiRNA binding sites shared by circRNAs")
                                        )
                
                #print(canb_heatmap)
                
                
                # Save at png
                ggsave(canb_heatmap, file=paste0(output_dir, "Heatmap_number_miRNA_binding_sites_per_circRNA.png"), width=w, height=h, dpi=600,  limitsize = FALSE, type="cairo")
            
            }
        }
        
    }

}

```

```{r, echo=FALSE, include=FALSE}

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    if (dim(bar1b)[1] > 0) {
    
        dt6 <- tibble::rownames_to_column(test, "miRNA_id")
        dt6 <- melt(dt6, id.vars = "miRNA_id", variable.name = "gene_circ_names", value.name = "count")
        dt6 <- as.data.table(dt6)
        dt6 <- dt6[count>0,]
        
        dt6$circ_id <- gsub("^.*_", "", dt6$gene_circ_names)
        
        #dt6$gene_circ_names <- paste0("<a href=\"", "/data/graphical_output/",dt6$circ_id, "/functional_predictions_", dt6$circ_id, ".html", "\" target=\"_blank\">", dt6$gene_circ_names, "</a>")
        
        dt6 <- dt6[, .(list(as.character(gene_circ_names))), by = miRNA_id]
        dt6$V1 <- gsub("^c", "", dt6$V1)
        dt6$V1 <- gsub("\"", "", dt6$V1)
        colnames(dt6) <- c("miRNA_id", "gene_circ_names")
    
    }

}

```

```{r pre_heatmap_density, echo=FALSE, include=FALSE}

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    if (dim(bar1c)[1] > 0) {
    
        density_miRNA_per_circRNA <- as.data.table(dcast(bar1c, miRNA_id ~ circ_id, value.var = "density", fill = 0))
    
    }

}

```

```{r table_miRNA_density, echo=FALSE, include=TRUE}

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    setkey(bar1c, circ_id)
    setkey(gene_names, circ_id)
    bar1c <- bar1c[gene_names, nomatch = 0]
    
    dt1 <- data.table(bar1c[, c(1,6,2,3,4)], round(bar1c[, 5], 2))
    
    dt1$gene_names <- paste0("<a href=\"", "https://www.genecards.org/cgi-bin/carddisp.pl?gene=", dt1$gene_names, "\" target=\"_blank\">", dt1$gene_names, "</a>")
    dt1$miRNA_id <- paste0("<a href=\"", "http://www.mirbase.org/textsearch.shtml?q=", dt1$miRNA_id, "&submit=submit", "\" target=\"_blank\">", dt1$miRNA_id, "</a>")

    
    datatable(data = dt1,
          rownames = F,
          colnames = c("CircRNA", "Gene name", "MiRNA", "Number of miRNA binding sites", "CircRNA length", "Density"),
          style = "bootstrap",
          class = "compact display",
          caption = paste0("CircRNAs with the correspondent number of miRNAs identified. CircRNA host gene name is linked to GeneCards, miRNA name is linked to miRBase. \"Density\" = ratio of the number of miRNA binding sites in a circRNA to circRNA length. Colored barplots highlight values in \"Number of miRNA binding sites\", \"CircRNA length\" and \"Density\" columns."),
          fillContainer = F,
          autoHideNavigation = T,
          filter = "top",
          escape = FALSE,
          extensions = c('ColReorder', 'Buttons', 'KeyTable'),
          options = list(colReorder = TRUE,
                         searching = TRUE,
                         pageLength = 10,
                         dom = 'Bfrtip',
                         buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                         keys = TRUE
                    )
    ) %>%
    formatStyle('num_siti',
                background = styleColorBar(dt1$num_siti, 'mediumorchid'),
                backgroundColor = "lightyellow",
                backgroundSize = '100% 90%',
                backgroundRepeat = 'no-repeat',
                backgroundPosition = 'center'
    ) %>%
    formatStyle('length',
                background = styleColorBar(dt1$length, 'steelblue'),
                backgroundColor = "lightyellow",
                backgroundSize = '100% 90%',
                backgroundRepeat = 'no-repeat',
                backgroundPosition = 'center'
    ) %>%
    formatStyle('density',
                background = styleColorBar(dt1$density, 'yellow'),
                backgroundColor = "lightyellow",
                backgroundSize = '100% 90%',
                backgroundRepeat = 'no-repeat',
                backgroundPosition = 'center'
    )

}

```

```{r table_miRNA_pos, echo=FALSE, include=TRUE}

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    dt2 <- miRNA_pred
    
    dt2$miRNA_id <- paste0("<a href=\"", "http://www.mirbase.org/textsearch.shtml?q=", dt2$miRNA_id, "&submit=submit", "\" target=\"_blank\">", dt2$miRNA_id, "</a>")
    dt2$gene_names <- paste0("<a href=\"", "https://www.genecards.org/cgi-bin/carddisp.pl?gene=", dt2$gene_names, "\" target=\"_blank\">", dt2$gene_names, "</a>")
    
    dt2 <- dt2[, c(1,5,2,3,4)]

    
    datatable(data = dt2,
          rownames = F,
          colnames = c("CircRNA", "Gene name", "MRE", "MRE start", "MiRNA end"),
          style = "bootstrap",
          class = "compact display",
          caption = paste0("MiRNAs identified with respective binding position. MRE: miRNA recognition element. CircRNA host gene name is linked to GeneCards, miRNA name is linked to miRBase. \"MRE start\" = MRE start position estimated from \"MRE end\" position for a 22 nt miRNA binding. MRE start and end positions are relative to circRNA sequence."),
          fillContainer = F,
          autoHideNavigation = T,
          filter = "top",
          escape = FALSE,
          extensions = c('ColReorder', 'Buttons', 'KeyTable'),
          options = list(colReorder = TRUE,
                         searching = TRUE,
                         pageLength = 10,
                         dom = 'Bfrtip',
                         buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                         keys = TRUE
                    )
        )

}

```

```{r table_common_miRNA, echo=FALSE, include=TRUE}

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    if (dim(bar1b)[1] > 0) {
        
        dt6$miRNA_id <- paste0("<a href=\"", "http://www.mirbase.org/textsearch.shtml?q=", dt6$miRNA_id, "&submit=submit", "\" target=\"_blank\">", dt6$miRNA_id, "</a>")
    
        datatable(data = dt6,
              rownames = F,
              colnames = c("MiRNA", "CircRNA"),
              style = "bootstrap",
              class = "compact display",
              #caption = paste0("MiRNAs shared by circRNAs. MiRNA name is linked to miRBase, circRNA name is linked to functional_predictions_circNAME.html."),
              caption = paste0("MiRNAs shared by circRNAs. MiRNA name is linked to miRBase."),
              fillContainer = F,
              autoHideNavigation = T,
              filter = "top",
              escape = FALSE,
              extensions = c('ColReorder', 'Buttons', 'KeyTable'),
              options = list(colReorder = TRUE,
                             searching = TRUE,
                             pageLength = 10,
                             dom = 'Bfrtip',
                             buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                             keys = TRUE
                        )
            )
    
    }

}

```

```{r print_2, echo=FALSE, include=FALSE}

print_2 <- ""

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    if (dim(bar1b)[1] > 0) {
        
        if (dim(test)[1]>=2 & dim(test)[1]<=500 & dim(test)[2]>=2 & dim(test)[2]<=30) {
            
            if (out == 0) {
            
                print_2 <- "The heatmap of common miRNAs (shared by at least two circRNAs) is saved in \"Heatmap_number_miRNA_binding_sites_per_circRNA.png\"."
                
            }
            
        }
        
    }
    
}

```

`r print_2`

```{r pre_heatmap_example_miRNA, echo=FALSE, include=FALSE}

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    if (dim(bar1b)[1] > 0) {
        
        if (dim(test)[1]>=2 & dim(test)[1]<=500 & dim(test)[2]>=2 & dim(test)[2]<=30) {
            
            if (out == 0) {
            
                test_1 <- matrix(c(0,1,4, 0,3,2, 2,2,0, 1,0,3), byrow = T, ncol = 3)
                rownames(test_1) <- c("miRNA 1", "miRNA 2", "miRNA 3", "miRNA 4")
                
                colnames(test_1) <- c("circRNA 1", "circRNA 2", "circRNA 3")
                
            }
            
        }
        
    }
    
}

```

```{r heatmap_example_miRNA, echo=FALSE, include=TRUE, width=4, height=4}

# heatmap

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    if (dim(bar1b)[1] > 0) {
        
        if (dim(test)[1]>=2 & dim(test)[1]<=500 & dim(test)[2]>=2 & dim(test)[2]<=30) {
            
            if (out == 0) {
                
                dist = "canberra"
                
                canb_heatmap <- pheatmap(mat = test_1,
                                         clustering_distance_rows = dist,
                                         clustering_distance_cols = dist,
                                         
                                         color = colorRampPalette(c("white","#1565C0"))(1000),
                
                                         cellheight = 10, cellwidth = 20, border_color="gray",
                                         
                                         treeheight_row = 20, treeheight_col = 20,
                                         
                                         main = "Example heatmap"
                                        )
                
                print(canb_heatmap)
                
            }

        }
        
    }
    
}

```



## Validated target genes (TGs)

```{r validated_TG, echo=FALSE, include=FALSE}

# validated target genes

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    if (parameters[2] == "hsa" | parameters[2] == "mmu" | parameters[2] == "rno") {

        ## multiMir

        if (dim(bar1b)[1] > 0) {

            multimir_results_all <- data.table()

            for (circ in circ_length$circ_id) {

                miRNA_id <- bar1b[circ_id == circ,]$miRNA_id

                if (length(miRNA_id) > 0) {

                    specie <- case_when(
                                 parameters[2] == "hsa" ~ "hsa",
                                 parameters[2] == "mmu" ~ "mmu",
                                 parameters[2] == "rno" ~ "rno"
                                    )

                    multimir_results <- get_multimir(org = specie, mirna = miRNA_id, table = 'validated', summary = TRUE)
                    multimir_results <- as.data.table(multimir_results@data)

                    #filter for strong categories
                    multimir_results <- multimir_results[grep("PAR-CLIP|HITS-CLIP|CLASH|Luciferase|Degradome|ChIP-seq|ELISA|Immuno.*", experiment),]
                    multimir_results <- multimir_results[support_type != "Functional MTI (Weak)",]

                    if (multimir_results[, .N] != 0) {

                        multimir_results <- data.table("circ_id" = circ, "miRNA_id" = multimir_results$mature_mirna_id, "target_gene" = multimir_results$target_symbol)
                        multimir_results_all <- rbind(multimir_results_all, multimir_results)

                    }

                }

            }

            if (multimir_results_all[, .N] > 0) {

                multimir_results_all <- unique(multimir_results_all)
                multimir_results_all <- multimir_results_all[target_gene != "",]

                setkey(multimir_results_all, circ_id)
                setkey(gene_names, circ_id)
                multimir_results_all <- multimir_results_all[gene_names, nomatch = 0]

                multimir_results_all <- multimir_results_all[, c(1,4,2,3)]

                # Save table
                write.table(multimir_results_all, paste0(output_dir, "All_validated_TGs.csv"), sep = ",", row.names = F, quote = F, dec = ",")

            }

        }

    }

}

```

```{r print_3, echo=FALSE, include=FALSE}

print_3 <- "Number of identified TGs: 0."

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {
    
    if (parameters[2] == "hsa" | parameters[2] == "mmu" | parameters[2] == "rno") {

        if (dim(bar1b)[1] > 0) {
            
            if (multimir_results_all[, .N] > 0) {
                
                print_3 <- "Network of common TGs shared by at least two circRNAs."
                
            }
        
        }
    
    }
    
}

```

`r print_3`

```{r print_3b, echo=FALSE, include=FALSE}

print_3b <- ""

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {
    
    if (parameters[2] == "hsa" | parameters[2] == "mmu" | parameters[2] == "rno") {

        if (dim(bar1b)[1] > 0) {
            
            if (multimir_results_all[, .N] > 0) {
                
                print_3b <- "Tip: if the network is too crowded, use more stringent parameters."
                
            }
        
        }
    
    }
    
}

```

`r print_3b`

```{r network_TG, echo=FALSE, include=TRUE, fig.width=12, fig.height=9}

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    if (parameters[2] == "hsa" | parameters[2] == "mmu" | parameters[2] == "rno") {

        if (dim(bar1b)[1] > 0) {

            if (multimir_results_all[, .N] > 0) {

                multimir_results_all$gene_circ_names <- paste0("circ", multimir_results_all$gene_names, "_", multimir_results_all$circ_id)

                # select only TGs common to alt 2 circRNAs
                dt0 <- multimir_results_all[, .(list(unique(gene_circ_names)), length(unique(gene_circ_names))), by = target_gene]
                colnames(dt0) <- c("target_gene", "gene_circ_names", "N")
                common_TGs <- dt0[N >= 2, "target_gene"]
                multimir_results_all <- multimir_results_all[target_gene %in% common_TGs$target_gene,]

                if (multimir_results_all[, .N] > 0) {

                    dt1 <- multimir_results_all[, .N, by = .(gene_circ_names, miRNA_id)]
                    dt2 <- multimir_results_all[, .N, by = .(miRNA_id, target_gene)]
                    colnames(dt1) <- c("term1", "term2", "freq")
                    colnames(dt2) <- c("term1", "term2", "freq")
                    dt <- rbind(dt1, dt2)

                    dt <- as.data.frame(dt)
                    gr <- graph.data.frame(dt[dt$freq!=0,])

                    a <- length(unique(multimir_results_all$gene_circ_names))
                    b <- length(unique(multimir_results_all$miRNA_id))
                    c <- length(unique(multimir_results_all$target_gene))

                    # default version
                    V(gr)$type <- c(rep("circRNA", a), rep("miRNA", b), rep("TG", c))
                    ty <- match(V(gr)$type, c("circRNA", "miRNA", "TG") )

                    #p1 <- plot(gr, vertex.color=ty)

                    # custom version
                    co <- c(rep("#dd6b6b", a), rep("#3ba6df", b), rep("#4bb815", c))

                    sh <- c(rep("circle", a), rep("rectangle", b), rep("square", c))

                    vs <- c(rep(15, a), rep(50, b), rep(20, c))

                    el <- c(rep("solid", dim(dt1)[1]), rep("longdash", dim(dt2)[1]))

                    p1 <- plot(gr,
                               vertex.color=co,
                               vertex.shape=sh,
                               vertex.size=vs,                                         # l
                               vertex.size2=10,                                        # h
                               vertex.label.family="Arial",
                               vertex.label.cex=1,
                               edge.lty=el,
                               arrow.mode=1,
                               margin=0
                               )

                }

            }

        }

    }

}

```

```{r table_common_TG, echo=FALSE, include=TRUE}

if (parameters[1] == "M" | parameters[1] == "MR" | parameters[1] == "MO" | parameters[1] == "MRO") {

    if (parameters[2] == "hsa" | parameters[2] == "mmu" | parameters[2] == "rno") {

        if (dim(bar1b)[1] > 0) {

            if (multimir_results_all[, .N] > 0) {

                dt8 <- multimir_results_all[, -"gene_circ_names"]

                dt8$gene_names <- paste0("<a href=\"", "https://www.genecards.org/cgi-bin/carddisp.pl?gene=", dt8$gene_names, "\" target=\"_blank\">", dt8$gene_names, "</a>")
                dt8$miRNA_id <- paste0("<a href=\"", "http://www.mirbase.org/textsearch.shtml?q=", dt8$miRNA_id, "&submit=submit", "\" target=\"_blank\">", dt8$miRNA_id, "</a>")
                dt8$target_gene <- paste0("<a href=\"", "https://www.genecards.org/cgi-bin/carddisp.pl?gene=", dt8$target_gene, "\" target=\"_blank\">", dt8$target_gene, "</a>")


                datatable(data = dt8,
                      rownames = F,
                      colnames = c("CircRNA", "Gene name", "MiRNA", "Target gene"),
                      style = "bootstrap",
                      class = "compact display",
                      caption = paste0("Target genes shared by at least two circRNAs. CircRNA host gene and target gene names are linked to GeneCards, miRNA name is linked to miRBase."),
                      fillContainer = F,
                      autoHideNavigation = T,
                      filter = "top",
                      escape = FALSE,
                      extensions = c('ColReorder', 'Buttons', 'KeyTable'),
                      options = list(colReorder = TRUE,
                                     searching = TRUE,
                                     pageLength = 10,
                                     dom = 'Bfrtip',
                                     buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                                     keys = TRUE
                                )
                    )

            }

        }

    }

}

```





# RNA binding protein (RBP) binding sites prediction

```{r print_4, echo=FALSE, include=FALSE}

print_4 <- "Analysis not performed."

if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    print_4 <- ""

}

```

`r print_4`

## Method statistics

```{r, echo=FALSE, include=FALSE}

## RBP predictions

if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    # beRBP
    
    # remove circRNAs with length > "l"
    beRBP_pred_l <- beRBP_pred[circ_id %in% circ_l,]
    
    
    # filter on "voteFrac"
    if (params$QUANTILE3 == "TRUE") {
        voteFrac_RBP <- quantile(beRBP_pred_l$voteFrac, params$thr3)
    } else {
        voteFrac_RBP <- params$voteFrac_RBP
    }
    
    beRBP_pred_l_sel <- beRBP_pred_l[(voteFrac >= voteFrac_RBP),]

}

```

```{r, echo=FALSE, include=FALSE}

if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    beRBP_pred_l_sel$circ_RBP_pos_id <- paste0(beRBP_pred_l_sel$circ_id, "__", beRBP_pred_l_sel$RBP, "__", beRBP_pred_l_sel$sitePos, "__", beRBP_pred_l_sel$siteLen)
    
    v <- unique(beRBP_pred_l_sel$circ_RBP_pos_id)
    RBP_pred <- data.table(t(as.data.table(strsplit(v, "__"))))
    colnames(RBP_pred) <- c("circ_id", "RBP", "sitePos", "siteLen")

}

```

```{r, echo=FALSE, include=FALSE}

if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    setkey(RBP_pred, circ_id)
    setkey(gene_names, circ_id)
    RBP_pred <- RBP_pred[gene_names, nomatch = 0]
    
    RBP_pred$gene_circ_names <- paste0("circ", RBP_pred$gene_names, "_", RBP_pred$circ_id)

} else {
    
    voteFrac_RBP <- params$voteFrac_RBP
    beRBP_pred_l_sel <- data.table()
    
}

```

```{r print_5, echo=FALSE, include=FALSE}

print_5 <- ""

if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    print_5 <- paste0("beRBP threshold: voteFrac > ", voteFrac_RBP)

}

```

`r print_5`

|                                      | beRBP                      |
|--------------------------------------|--------------------------- |
| Number of RBP binding over threshold | `r beRBP_pred_l_sel[, .N]` |


## RBP binding sites per circRNA

```{r pre_barplot_1_RBP, echo=FALSE, include=FALSE}

if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    bar2a <- RBP_pred[, .N, by = gene_circ_names]
    
    w <- bar2a[, .N]
    w <-w/2
    if (w < 5) {
        w <- 5
    }

}

```

```{r barplot_1_RBP, echo=FALSE, include=TRUE}

if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    # barplot of the total number of detected RBPs in each circRNA
    
    bar_plot2 <- ggplot(bar2a,
                       aes(x = gene_circ_names, y = N, fill = gene_circ_names)
                      ) +
                geom_bar(stat = "identity") +
                geom_text(aes(label = N),
                          vjust = -0.3,
                          size = 3.5) +
                scale_fill_viridis(discrete = TRUE, option = "C") +
                xlab("circRNA") +
                ylab("Number of RBP binding sites") +
                ggtitle(paste0("Overall number of RBP binding sites per circRNA")) +
                theme_bw() +
                theme(plot.title = element_text(hjust = 0.5),
                      text = element_text(size=10),
                      axis.text.x = element_text(angle=70, hjust=1),
                      legend.position = "none")
    
    print(bar_plot2)
    
    
    # Save at png
    ggsave(bar_plot2, file=paste0(output_dir, "Overall_number_of_RBP_binding_sites_per_circRNA.png"), width=w, height=7, dpi=600, limitsize = TRUE, type = "cairo")

}

```

```{r datatable_circ_RBP_length, echo=FALSE, include=TRUE}

if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    bar2a$gene_names <- gsub("circ", "", bar2a$gene_circ_names)
    bar2a$gene_names <- gsub("_.*$", "", bar2a$gene_names)
    bar2a$circ_id <- gsub("^.*_", "", bar2a$gene_circ_names)
    
    bar2a2 <- RBP_pred[, .(length(unique(RBP))), by = gene_circ_names]
    setkey(bar2a, gene_circ_names)
    setkey(bar2a2, gene_circ_names)
    bar2a <- bar2a[bar2a2, nomatch = 0]
    colnames(bar2a)[5] <- "N_unique"
    
    setkey(bar2a, circ_id)
    setkey(circ_length, circ_id)
    table2a <- merge(bar2a, circ_length)
    table2a <- table2a[, c(1,4,3,5,6)]
    
    datatable(data = table2a,
              rownames = F,
              colnames = c("CircRNA", "Gene name", "Number of RBP binding sites", "Number of RBPs", "CircRNA length"),
              style = "bootstrap",
              class = "compact display",
              fillContainer = F,
              autoHideNavigation = T,
              filter = "top",
              escape = FALSE,
              extensions = c('ColReorder', 'Buttons', 'KeyTable'),
              options = list(colReorder = TRUE,
                             searching = TRUE,
                             pageLength = 10,
                             dom = 'Bfrtip',
                             buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                             keys = TRUE
                            )
            )

}

```

```{r density_RBP, echo=FALSE, include=FALSE}

if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    bar2b <- RBP_pred[, .N, by = .(circ_id, RBP)]
    
    setkey(bar2b, circ_id)
    setkey(circ_length, circ_id)
    bar2c <- bar2b[circ_length, nomatch = 0]
    
    colnames(bar2c)[3] <- "num_siti"
    
    bar2c <- data.table(bar2c, "density" = bar2c$num_siti/bar2c$length*100)

} else {
    
    RBP_pred <- data.table(RBP = integer())
    
}

```

Number of different identified RBPs: `r length(unique(RBP_pred$RBP))`.


```{r pre_heatmap_num_RBP, echo=FALSE, include=FALSE}

if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    setkey(bar2b, circ_id)
    setkey(gene_names, circ_id)
    bar2b <- bar2b[gene_names, nomatch = 0]
    
    bar2b$gene_circ_names <- paste0("circ", bar2b$gene_names, "_", bar2b$circ_id)

    if (dim(bar2b)[1] > 0) {
        
        num_RBP_per_circRNA <- as.data.table(dcast(bar2b, RBP ~ gene_circ_names, value.var = "N", fill = 0))

    }
    
    # pre heatmaps
    
    # input
    
    if (dim(bar2b)[1] > 0) {
        
        test <- num_RBP_per_circRNA[, !"RBP"]
        rownames(test) <- num_RBP_per_circRNA$RBP
        
        #print only common RBPs
        test$fun <- apply(test, 1, function(x) length(which(as.integer(x)>0)))
        test <- tibble::rownames_to_column(test, "RBP")
        test <- test[fun >= 2,]
        
        test <- test[, -"fun"]
        test <- tibble::column_to_rownames(test, "RBP")
        
        
        w <- dim(test)[2]/1.67
        if (w < 5) {
            w <- 5
        }
        h <- dim(test)[1]/5
        if (h < 7) {
            h <- 7
        }
    
    }

}

```

```{r canberra_heatmap_num_RBP, echo=FALSE, include=FALSE}

if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    # heatmap
    
    if (dim(bar2b)[1] > 0) {
        
        if (dim(test)[1]>=2 & dim(test)[1]<=500 & dim(test)[2]>=2 & dim(test)[2]<=30) {

            out <- 0
            i <- 1
            
            while (out == 0) {
                brks <- quantile(test[,i])
                if (length(brks) != length(unique(brks))) {
                    out <- 1
                }
                i <- i+1
            }
            
            if (out == 0) {
                
                dist = "canberra"
            
                canb_heatmap <- pheatmap(mat = test,
                              clustering_distance_rows = dist,
                              clustering_distance_cols = dist,
                
                              color = colorRampPalette(c("white","#E65100"))(1000),
                
                              cellheight = 10, cellwidth = 20, border_color="gray",
                              
                              treeheight_row = 20, treeheight_col = 20,
    
                              main = paste0("RBP binding sites shared by circRNAs")
                             )
                
                #print(canb_heatmap)
                
                
                # Save at png
                ggsave(canb_heatmap, file=paste0(output_dir, "Heatmap_number_RBP_binding_sites_per_circRNA.png"), width=w, height=h, dpi=600, limitsize = FALSE, type="cairo")
        
            }
            
        }
    
    }

}

```

```{r, echo=FALSE, include=FALSE}

if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    if (dim(bar2b)[1] > 0) {
        
        dt7 <- tibble::rownames_to_column(test, "RBP")
        dt7 <- melt(dt7, id.vars = "RBP", variable.name = "gene_circ_names", value.name = "count")
        dt7 <- as.data.table(dt7)
        dt7 <- dt7[count>0,]
        
        dt7$circ_id <- gsub("^.*_", "", dt7$gene_circ_names)
        
        #dt7$gene_circ_names <- paste0("<a href=\"", dt7$circ_id, "/functional_predictions_", dt7$circ_id, ".html", "\" target=\"_blank\">", dt7$gene_circ_names, "</a>")
        
        dt7 <- dt7[, .(list(as.character(gene_circ_names))), by = RBP]
        dt7$V1 <- gsub("^c", "", dt7$V1)
        dt7$V1 <- gsub("\"", "", dt7$V1)
        colnames(dt7) <- c("RBP", "gene_circ_names")
    
    }

}

```

```{r pre_heatmap_density_RBP, echo=FALSE, include=FALSE}

if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    if (dim(bar2c)[1] > 0) {
        
        density_RBP_per_circRNA <- as.data.table(dcast(bar2c, RBP ~ circ_id, value.var = "density", fill = 0))
    
    }
    
}

```

```{r table_RBP_density, echo=FALSE, include=TRUE}

if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    setkey(bar2c, circ_id)
    setkey(gene_names, circ_id)
    bar2c <- bar2c[gene_names, nomatch = 0]
    
    #bar1c$gene_circ_names <- paste0("circ", bar1c$gene_names, "_", bar1c$circ_id)
    
    dt3 <- data.table(bar2c[, c(1,6,2,3,4)], round(bar2c[, 5], 2))
    
    require(params$orgdb, character.only = TRUE)
    tab <- toTable(eval(parse(text = params$symbol2eg)))   # tab <- toTable(params$symbol2eg)
    tab <- as.data.table(tab)
    
    colnames(dt3)[3] <- "symbol"
    
    setkey(dt3, symbol)
    setkey(tab, symbol)
    dt3 <- merge(dt3, tab, all.x=T)
    
    tabU <- toTable(eval(parse(text = params$eg2uniprot)))   # tabU <- toTable(params$eg2uniprot)
    tabU <- as.data.table(tabU)
    tabU$uniprot_id <- paste0("<a href=\"", "https://www.uniprot.org/uniprot/", tabU$uniprot_id, "\" target=\"_blank\">", tabU$uniprot_id, "</a>")
    tabU <- tabU[, .(list(uniprot_id)), by = gene_id]
    colnames(tabU)[2] <- "uniprot_id"
    
    setkey(dt3, gene_id)
    setkey(tabU, gene_id)
    dt3 <- merge(dt3, tabU, all.x=T)
    
    dt3 <- dt3[, -"gene_id"]
    colnames(dt3)[1] <- "RBP"
    
    dt3 <- dt3[, c(2,3,1,4,5,6,7)]
    
    dt3$gene_names <- paste0("<a href=\"", "https://www.genecards.org/cgi-bin/carddisp.pl?gene=", dt3$gene_names, "\" target=\"_blank\">", dt3$gene_names, "</a>")


    datatable(data = dt3,
              rownames = F,
              colnames = c("CircRNA", "Gene name", "RBP","Number of RBP binding sites", "CircRNA length", "Density", "Uniprot"),
              style = "bootstrap",
              class = "compact display",
              caption = paste0("CircRNA with the correspondent number of RBPs identified. CircRNA host gene name is linked to GeneCards, RBP is linked to UniProt. \"Density\" = ratio of the number of RBP binding sites in a circRNA to circRNA length. Colored barplots highlight values in \"Number of RB binding sites\", \"CircRNA length\" and \"Density\" columns."),
              fillContainer = F,
              autoHideNavigation = T,
              filter = "top",
              escape = FALSE, 
              extensions = c('ColReorder', 'Buttons', 'KeyTable'),
              options = list(colReorder = TRUE,
                             searching = TRUE,
                             pageLength = 10,
                             dom = 'Bfrtip',
                             buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                             keys = TRUE
                            )
            ) %>%
            formatStyle('num_siti',
                    background = styleColorBar(dt3$num_siti, 'mediumorchid'),
                    backgroundColor = "lightyellow",
                    backgroundSize = '100% 90%',
                    backgroundRepeat = 'no-repeat',
                    backgroundPosition = 'center'
                    ) %>%
            formatStyle('length',
                    background = styleColorBar(dt3$length, 'steelblue'),
                    backgroundColor = "lightyellow",
                    backgroundSize = '100% 90%',
                    backgroundRepeat = 'no-repeat',
                    backgroundPosition = 'center'
                    ) %>%
            formatStyle('density',
                    background = styleColorBar(dt3$density, 'yellow'),
                    backgroundColor = "lightyellow",
                    backgroundSize = '100% 90%',
                    backgroundRepeat = 'no-repeat',
                    backgroundPosition = 'center'
                    )
    
}

```

```{r table_RBP_pos, echo=FALSE, include=TRUE}

if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    dt4 <- RBP_pred
    
    colnames(dt4)[2] <- "symbol"
    
    setkey(dt4, symbol)
    setkey(tab, symbol)
    dt4 <- merge(dt4, tab, all.x=T)
    
    setkey(dt4, gene_id)
    setkey(tabU, gene_id)
    dt4 <- merge(dt4, tabU, all.x=T)
    
    dt4 <- dt4[, -"gene_id"]
    colnames(dt4)[1] <- "RBP"
    
    dt4 <- dt4[, c(2,5,1,3,4,7)]
    
    dt4$gene_names <- paste0("<a href=\"", "https://www.genecards.org/cgi-bin/carddisp.pl?gene=", dt4$gene_names, "\" target=\"_blank\">", dt4$gene_names, "</a>")


    datatable(data = dt4,
              rownames = F,
              colnames = c("CircRNA", "Gene name", "RBP", "RRE start", "RRE length", "Uniprot"),
              style = "bootstrap",
              class = "compact display",
              caption = paste0("RBPs identified with respective binding position. RRE: RBP recognition element. CircRNA host gene name is linked to GeneCards, RBP is linked to UniProt. \"RRE start\" position is relative to circRNA sequence."),
              fillContainer = F,
              autoHideNavigation = T,
              filter = "top",
              escape = FALSE,
              extensions = c('ColReorder', 'Buttons', 'KeyTable'),
              options = list(colReorder = TRUE,
                             searching = TRUE,
                             pageLength = 10,
                             dom = 'Bfrtip',
                             buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                             keys = TRUE
                            )
            )
    
}

```

```{r table_common_RBP, echo=FALSE, include=TRUE}

if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    if (dim(bar2b)[1] > 0) {
        
        colnames(dt7)[1] <- "symbol"
        
        setkey(dt7, symbol)
        setkey(tab, symbol)
        dt7 <- merge(dt7, tab, all.x=T)
        
        setkey(dt7, gene_id)
        setkey(tabU, gene_id)
        dt7 <- merge(dt7, tabU, all.x=T)
        
        dt7 <- dt7[, -"gene_id"]
        colnames(dt7)[1] <- "RBP"
        

        datatable(data = dt7,
                  rownames = F,
                  colnames = c("RBP", "CircRNA", "Uniprot"),
                  style = "bootstrap",
                  class = "compact display",
                  #caption = paste0("RBPs shared by circRNAs. CircRNA name is linked to functional_predictions_circNAME.html, RBP is linked to UniProt."),
                  caption = paste0("RBPs shared by circRNAs. RBP is linked to UniProt."),
                  fillContainer = F,
                  autoHideNavigation = T,
                  filter = "top",
                  escape = FALSE,
                  extensions = c('ColReorder', 'Buttons', 'KeyTable'),
                  options = list(colReorder = TRUE,
                                 searching = TRUE,
                                 pageLength = 10,
                                 dom = 'Bfrtip',
                                 buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                                 keys = TRUE
                                )
                )
        
    }

}

```

```{r print_6, echo=FALSE, include=FALSE}

print_6 <- ""

if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    if (dim(bar2b)[1] > 0) {
        
        if (dim(test)[1]>=2 & dim(test)[1]<=500 & dim(test)[2]>=2 & dim(test)[2]<=30) {
            
            if (out == 0) {

            print_6 <- "The heatmap of common RBPs (shared by at least two circRNAs) is saved in \"Heatmap_number_RBP_binding_sites_per_circRNA.png\"."
            
            }
            
        }
        
    }
    
}

```

`r print_6`

```{r pre_heatmap_example_RBP, echo=FALSE, include=FALSE}

if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    if (dim(bar2b)[1] > 0) {
        
        if (dim(test)[1]>=2 & dim(test)[1]<=500 & dim(test)[2]>=2 & dim(test)[2]<=30) {

            if (out == 0) {
                
                test_1 <- matrix(c(0,1,4, 0,3,2, 2,2,0, 1,0,3), byrow = T, ncol = 3)
                rownames(test_1) <- c("RBP 1", "RBP 2", "RBP 3", "RBP 4")
                
                colnames(test_1) <- c("circRNA 1", "circRNA 2", "circRNA 3")
                
            }
            
        }
        
    }
    
}

```

```{r heatmap_example_RBP, echo=FALSE, include=TRUE, width=4, height=4}

# heatmap

if (parameters[1] == "R" | parameters[1] == "MR" | parameters[1] == "RO" | parameters[1] == "MRO") {

    if (dim(bar2b)[1] > 0) {
        
        if (dim(test)[1]>=2 & dim(test)[1]<=500 & dim(test)[2]>=2 & dim(test)[2]<=30) {
            
            if (out == 0) {
            
                dist = "canberra"
                
                canb_heatmap <- pheatmap(mat = test_1,
                                         clustering_distance_rows = dist,
                                         clustering_distance_cols = dist,
                                         
                                         color = colorRampPalette(c("white","#E65100"))(1000),
                
                                         cellheight = 10, cellwidth = 20, border_color="gray",
                                         
                                         treeheight_row = 20, treeheight_col = 20,
                                         
                                         main = "Example heatmap"
                                        )
                
                print(canb_heatmap)

            }
            
        }
        
    }
    
}

```





# Open Reading Frames (ORFs) prediction

```{r print_7, echo=FALSE, include=FALSE}

print_7 <- "Analysis not performed."

if (parameters[1] == "O" | parameters[1] == "MO" | parameters[1] == "RO" | parameters[1] == "MRO") {
    
    print_7 <- "Circular plots showing an example of __ORF with start and stop codon__ in the upper plot and an example of __ORF without stop codon (rolling ORF)__ in the lower plot. A 1000 nucleotides (nt) circRNA is used as example. The circRNA is depicted with a gray circular line, the backsplice junction (_bks_) is printed in black. \"_nt before bks_\": nucleotide sequence from 5'-start to the backsplice junction (in orange), \"_nt after bks_\": nucleotide sequence from the backsplice junction to 3'-end (in dark red). In the lower plot, the ORF is stopped to the second round, but it potentially goes around the circRNA multiple times."

}

```

`r print_7`

```{r circular_barplot_ORF_example1, echo=FALSE, include=TRUE, fig.width=5, fig.height=5}

if (parameters[1] == "O" | parameters[1] == "MO" | parameters[1] == "RO" | parameters[1] == "MRO") {
    
    # Create dataset
    l <- 1000
    
    orf_start <- 450
    orf_end <- 150
    
    
    # Prepare a data frame for grid (scales)
    grid_data <- data.table("start" = 0,
                            "end" = l)
    
    # Make the plot
    circ_plot1 <- ggplot(grid_data, aes(x = start, y = 0)) +
        

                # Add lines
                geom_segment(data=grid_data, aes(x = start+4, y = 0, xend = end-4, yend = 0), colour = "darkgrey", alpha=1, size=2.5, inherit.aes = FALSE ) +
                geom_segment(data=grid_data, aes(x = end-4, y = 0, xend = end, yend = 0), colour = "black", alpha=1, size=2.5, inherit.aes = FALSE ) +
                geom_segment(data=grid_data, aes(x = start, y = 0, xend = start+4, yend = 0), colour = "black", alpha=1, size=2.5, inherit.aes = FALSE ) + 
        
                geom_segment(data=grid_data, aes(x = orf_start, y = 0.1, xend = l, yend = 0.15), colour = "orange", alpha=1, size=1, inherit.aes = FALSE ) + 
                geom_segment(data=grid_data, aes(x = 0, y = 0.15, xend = orf_end, yend = 0.2), colour = "darkred", alpha=1, size=1, inherit.aes = FALSE ) +
    
                # Add text
                annotate("text", x = c(0, 950), y = c(0.1, 1), label = c("0", "ORF with stop codon") , color=c("grey", "black"), size=c(3,4), angle=0, fontface=c("bold", "bold"), hjust=1) + 
                annotate("text", x = c(orf_start-20, orf_end+20), y = c(0.12, 0.22), label = c("5'", "3'"), color = c("orange", "darkred"), size = c(3,3), fontface = c("bold", "bold"), hjust=1) +
                annotate("text", x = 20, y = -0.12, label = "bks", color = "black", size = 3, fontface = "bold", hjust=1) +
                annotate("text", x = 650, y = 0.25, label = "nt before bks", color = "orange", size = 3.5, fontface = "bold", hjust=1) +
                annotate("text", x = 125, y = 0.75, label = "nt after bks", color = "darkred", size = 3.5, fontface = "bold", hjust=1) +
    
                ylim(-1, 1) +
        
                theme_minimal() +
                theme(
                    legend.position = "none",
                    axis.text = element_blank(),
                    text = element_text(size=8),
                    axis.title = element_blank(),
                    axis.text.x = element_text(angle=0, hjust=1)
                  ) +
    
                coord_polar(start = 0)

    
    print(circ_plot1)

}

```

```{r circular_barplot_ORF_example2, echo=FALSE, include=TRUE, fig.width=5, fig.height=5}

if (parameters[1] == "O" | parameters[1] == "MO" | parameters[1] == "RO" | parameters[1] == "MRO") {

    # Create dataset
    l <- 1000
    
    orf_start <- 450
    orf_end <- 1000
    
    
    # Prepare a data frame for grid (scales)
    grid_data <- data.table("start" = 0,
                            "end" = l)
    
    # Make the plot
    circ_plot2 <- ggplot(grid_data, aes(x = start, y = 0)) +
        
                # Add lines
                geom_segment(data=grid_data, aes(x = start+4, y = 0, xend = end-4, yend = 0), colour = "darkgrey", alpha=1, size=2.5, inherit.aes = FALSE ) +
                geom_segment(data=grid_data, aes(x = end-4, y = 0, xend = end, yend = 0), colour = "black", alpha=1, size=2.5, inherit.aes = FALSE ) +
                geom_segment(data=grid_data, aes(x = start, y = 0, xend = start+4, yend = 0), colour = "black", alpha=1, size=2.5, inherit.aes = FALSE ) + 
                
                geom_segment(data=grid_data, aes(x = orf_start, y = 0.1, xend = l, yend = 0.2), colour = "orange", alpha=1, size=1, inherit.aes = FALSE ) + 
                geom_segment(data=grid_data, aes(x = 0, y = 0.2, xend = orf_end, yend = 0.3), colour = "darkred", alpha=1, size=1, inherit.aes = FALSE ) +
    
                # Add text
                annotate("text", x = c(0, 970), y = c(0.1, 1), label = c("0", "ORF without stop codon") , color=c("grey", "black"), size=c(3,4), angle=0, fontface=c("bold", "bold"), hjust=1) + 
                annotate("text", x = c(orf_start-20, 20), y = c(0.12, 0.32), label = c("5'", "3'"), color = c("orange", "darkred"), size = c(3,3), fontface = c("bold", "bold"), hjust=1) +
                annotate("text", x = 20, y = -0.12, label = "bks", color = "black", size = 3, fontface = "bold", hjust=1) +
                annotate("text", x = 650, y = 0.35, label = "nt before bks", color = "orange", size = 3.5, fontface = "bold", hjust=1) +
                annotate("text", x = 125, y = 0.75, label = "nt after bks", color = "darkred", size = 3.5, fontface = "bold", hjust=1) +
        
                ylim(-1, 1) +
        
                theme_minimal() +
                theme(
                    legend.position = "none",
                    axis.text = element_blank(),
                    text = element_text(size=8),
                    axis.title = element_blank(),
                    axis.text.x = element_text(angle=0, hjust=1)
                  ) +
    
                coord_polar(start = 0)
    

    print(circ_plot2)

}

```

```{r, echo=FALSE, include=FALSE}

## ORF predictions

if (parameters[1] == "O" | parameters[1] == "MO" | parameters[1] == "RO" | parameters[1] == "MRO") {

    # ORFfinder
    
    # remove circRNAs with length > "l"
    ORFfinder_pred_l <- ORFfinder_pred[circ_id %in% circ_l,]
    
    openORF_l <- openORF[circ_id %in% circ_l,]
    openORF_l[open == "open",]$open <- "T"

}

```

```{r, echo=FALSE, include=TRUE}

if (parameters[1] == "O" | parameters[1] == "MO" | parameters[1] == "RO" | parameters[1] == "MRO") {

    # join close and open ORF
    
    #"ORFfinder_pred_l":
    if (ORFfinder_pred_l[, .N] > 0) {
    
        ORFfinder_pred_l <- ORFfinder_pred_l[, -"ORF"]
        
        setkey(ORFfinder_pred_l, circ_id)
        setkey(circ_length, circ_id)
        ORFfinder_pred_l <- ORFfinder_pred_l[circ_length, nomatch = 0]
        
        ORFfinder_pred_l$nt_before_bks <- ORFfinder_pred_l$length - ORFfinder_pred_l$start + 1
        ORFfinder_pred_l$nt_after_bks <- ORFfinder_pred_l$end - ORFfinder_pred_l$length
        ORFfinder_pred_l$open <- "F"
        
        ORFfinder_pred_l <- ORFfinder_pred_l[, c("circ_id", "length", "start", "end", "nt_before_bks", "nt_after_bks", "length_nt", "length_aa", "open")]
    
    }
    
    
    #"openORF_l":
    if (openORF_l[, .N] > 0) {
    
        openORF_l$length_nt <- openORF_l$end - openORF_l$start + 1
        openORF_l$length_aa <- openORF_l$length_nt / 3
        
        setkey(openORF_l, circ_id)
        setkey(circ_length, circ_id)
        openORF_l <- openORF_l[circ_length, nomatch = 0]
        
        openORF_l$nt_before_bks <- ifelse(openORF_l$start <= openORF_l$length, openORF_l$length - openORF_l$start+1, 2*openORF_l$length - openORF_l$start+1)
        openORF_l$nt_after_bks <- ifelse(openORF_l$start <= openORF_l$length, openORF_l$end - openORF_l$length, openORF_l$end - 2*openORF_l$length)
        
        openORF_l <- openORF_l[, c("circ_id", "length", "start", "end", "nt_before_bks", "nt_after_bks", "length_nt", "length_aa", "open")]
    
    }
    
    
    #"ORFfinder_pred_l" and "openORF_l":
    if ((ORFfinder_pred_l[, .N] > 0) & (openORF_l[, .N] > 0)) {
        
        ORF_pred <- rbind(ORFfinder_pred_l, openORF_l)
        ORF_pred$open <- factor(ORF_pred$open, levels = c("T", "F"))
    
    } else {
        
        if (ORFfinder_pred_l[, .N] > 0) {
            
            ORF_pred <- ORFfinder_pred_l
            ORF_pred$open <- factor(ORF_pred$open, levels = c("T", "F"))
            
        } else {
            
            if (openORF_l[, .N] > 0) {
            
                ORF_pred <- openORF_l
                ORF_pred$open <- factor(ORF_pred$open, levels = c("T", "F"))
            
            } else {
                
                ORF_pred <- data.table()
                print("No predicted ORFs")
                
            }
        }
    }

}

```

```{r pre_barplot_ORF, echo=FALSE, include=FALSE}

if (parameters[1] == "O" | parameters[1] == "MO" | parameters[1] == "RO" | parameters[1] == "MRO") {

    if (ORF_pred[, .N] > 0) {
        
        setkey(ORF_pred, circ_id)
        setkey(gene_names, circ_id)
        ORF_pred <- ORF_pred[gene_names, nomatch=0]
        
        ORF_pred$gene_circ_names <- paste0("circ", ORF_pred$gene_names, "_", ORF_pred$circ_id)

        ORF_pred[open == "F",]$open <- "with stop codon"
        ORF_pred[open == "T",]$open <- "without stop codon"
        
        dt <- as.data.table(dcast(ORF_pred, gene_circ_names ~ open, fun.aggregate = length))
        bar3a <- melt(dt, id.vars = "gene_circ_names", variable.name = "open", value.name = "count")

        w <- bar3a[, .N]
        w <- max(w,2)
        w <- w/3*2
        if (w < 8) {
            w <- 8
        }
        
        b <- max(ORF_pred[, .N, by = .(circ_id, open)]$N)
        h <- b + 4
        if (h < 6) {
            h <- 6
        }
    
    }

}

```

```{r barplot_1_ORF, echo=FALSE, include=TRUE}

if (parameters[1] == "O" | parameters[1] == "MO" | parameters[1] == "RO" | parameters[1] == "MRO") {

    # barplot of the total number of detected ORFs in each circRNA
    
    if (ORF_pred[, .N] > 0) {
        
        bar_plot7 <- ggplot(bar3a,
                           aes(x = gene_circ_names, y = count, fill = open)
                          ) +
                    geom_bar(stat = "identity", position=position_dodge()) + 
                    scale_fill_manual(values = c("lightblue","gray")) +
                    scale_y_continuous(breaks=seq(0,b,1)) +
                    xlab("CircRNA") +
                    ylab("Number of detected ORFs") +
                    labs(fill = "ORF") +
                    ggtitle(paste0("Overall number of ORFs per circRNA")) +
                    theme_bw() +
                    theme(plot.title = element_text(hjust = 0.5),
                          text = element_text(size=16),
                          axis.text.x = element_text(angle=70, hjust=1, size=14)
                          )

        #print(bar_plot7)
        
            
        # Save at png
        ggsave(bar_plot7, file=paste0(output_dir, "Overall_number_of_ORFs_per_circRNA.png"), width=w, height=h, limitsize = FALSE, type="cairo")
    
    }

}

```

```{r print_8, echo=FALSE, include=FALSE}

print_8 <- "Number of detected ORFs: 0."

if (parameters[1] == "O" | parameters[1] == "MO" | parameters[1] == "RO" | parameters[1] == "MRO") {
    
    if (ORF_pred[, .N] > 0) {
    
        print_8 <- "The barplot of detected ORFs per circRNA is saved in \"Overall_number_of_ORFs_per_circRNA.png\"."

    }
    
}

```

`r print_8`

```{r table_ORF, echo=FALSE, include=TRUE}

if (parameters[1] == "O" | parameters[1] == "MO" | parameters[1] == "RO" | parameters[1] == "MRO") {

    if (ORF_pred[, .N] > 0) {
    
        ORF_pred$aa_before_bks <- round(ORF_pred$nt_before_bks /3, 1)
        ORF_pred$aa_after_bks <- round(ORF_pred$nt_after_bks /3, 1)
        ORF_pred$inframe <- ifelse(ORF_pred$nt_before_bks%%3 == 0, "T", "F")
        
        dt5 <- ORF_pred[, c(1,10,7,8,9,14,2,3,4,5,12,6,13)]   # change column order
        
        dt5$gene_names <- paste0("<a href=\"", "https://www.genecards.org/cgi-bin/carddisp.pl?gene=", dt5$gene_names, "\" target=\"_blank\">", dt5$gene_names, "</a>")
    
    } else {
    
        # print empty datatable
        dt5 <- data.table("circ_id" = character(), "length_nt" = numeric(), "length_aa" = numeric(), "open" = character(), "length" = numeric(), "start" = numeric(), "end" = numeric(), "nt_before_bks" = numeric(), "nt_after_bks" = numeric())
        
    }
    
    
    table5 <- datatable(data = dt5,
          rownames = F,
          colnames = c("CircRNA", "Gene name", "ORF length (nt)", "ORF length (aa)", "ORF", "Inframe (T/F)", "CircRNA length (nt)", "ORF start", "ORF end", "nt before bks", "aa before bks", "nt after bks", "aa after bks"),
          style = "bootstrap",
          class = "compact display",
          caption = paste0("Predicted Open Reading Frames crossing the backsplice junction (bks). \"ORF\", \"nt before bks\" and \"nt after bks\": as detailed before; \"aa before bks\" and \"aa after bks\": amino acid sequence correspondent to \"nt before bks\" and \"nt after bks\", respectively. \"ORF start\" and \"ORF end\" positions are relative to circRNA sequence."),
          fillContainer = F,
          autoHideNavigation = T,
          filter = "top",
          escape = FALSE,
          extensions = c('ColReorder', 'Buttons', 'KeyTable'),
          options = list(colReorder = TRUE,
                         searching = TRUE,
                         pageLength = 10,
                         dom = 'Bfrtip',
                         buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                         keys = TRUE
                        )
        ) %>%
    formatStyle('open',
                target = "row",
                backgroundColor = styleEqual(c("without stop codon", "with stop codon"), c('lightgray', 'lightblue'))
                )
    
    
    table5
    
}

```





# Session info

This page was generated with the following packages version:

```{r session, echo=FALSE}

sessionInfo()

```


